src/Castle.Core/DynamicProxy/AbstractInvocation.cs

```diff
@@ -15,13 +15,17 @@
 namespace Castle.DynamicProxy
 {
 	using System;
+	using System.Collections.Generic;
 	using System.Diagnostics;
 	using System.Reflection;
 
+	using Castle.DynamicProxy.Generators.Emitters;
+
 	public abstract class AbstractInvocation : IInvocation
 	{
 		private readonly IInterceptor[] interceptors;
 		private readonly object[] arguments;
+		private readonly object[] originalArguments;
```

Comment:
This is the expensive bit. We end up keeping two copies of the whole argument list (when there are by-ref parameters). It would've been much more efficient to have a set of dirty flags, one per parameter. (See PR description above.)